### DISCUSSION TITLE
How can I obtain all device data when the device service is started.

### DISCUSSION THREAD

**Discussion by yangcheng999** [edgexfoundry/community#187] [Q&A] [answered]
http://{IP}:59881/api/v3/device/service/name/{deviceService}
----------------
During the device service startup process， device data will be obtained from core-metadata using above API.
But one time request can‘t obtain all device data.

Response Example :
-----------------------------------------------
{"apiVersion":"v3","statusCode":200,"totalCount":959,"devices":[***1**]
-----------------------------------------------
There are only 20 device data at position***1**.  
When there is a large amount of device data, it will not be possible to monitor all device.

**Comment by yangcheng999** [comment]
Using device-sdk-c.

**Reply by jiekechoo** [reply, marked as answer]
try this: https://github.com/edgexfoundry/device-sdk-c/blob/4c52de37db7d3bd0277fba2739b75d0a1bd475b3/include/devsdk/devsdk.h#L324

**Reply by yangcheng999** [reply]
I know this function. But this function obtain data from local cache(is not from core-metadata).  At this point, the service has been started And completed. Local cache only loaded 20 devices. This funcion also can only obtain 20 devices. 
During device service starting, it will obtain device data through network.  ->refer to : metadata.c -> function: edgex_metadata_client_get_devices

http://{IP}:59881/api/v3/device/service/name/{deviceService}  **?limit=-1**
If parameter 【 limit=-1】 are added, all device data can be obtained，But this will modify SDK.
like this：
---------------------------------
-```
  snprintf
  (
    url,
    URL_BUF_SIZE - 1,
    "http://%s:%u/api/" EDGEX_API_VERSION "/device/service/name/%s",
    endpoints->metadata.host,
    endpoints->metadata.port,
    ename
  );

  iot_data_t *jwt_data = edgex_secrets_request_jwt (secretprovider);  
  ctx.jwt_token = iot_data_string(jwt_data);
  //yangc add get all device data
  **strcat(url, "?limit=-1");**
```
----------------------------------

**Reply by cloudxxx8** [reply]
@FelixTing are you able to help this question?

**Reply by yangcheng999** [reply]
@FelixTing Do you have any better ideas？

**Reply by FelixTing** [reply]
@ych988 Your are right. The query parameter `limit=-1` is also used in device-sdk-go during the initialization of the device cache: https://github.com/edgexfoundry/device-sdk-go/blob/d1553bc24bec90db76a2b10fa9e928c5afbf1e1c/internal/cache/init.go#L27

**Comment by cloudxxx8** [comment]
close as answered.
please open an issue and PR to SDK if you want.

**Reply by FelixTing** [reply]
an issue was created for this
https://github.com/edgexfoundry/device-sdk-c/issues/510
